v-next/core/src/internal/providers/http.ts

```diff
@@ -0,0 +1,361 @@
+   * times out.
+   */
+  public async request(
+    requestArguments: RequestArguments,
+  ): Promise<SuccessfulJsonRpcResponse["result"]> {
+    const { method, params } = requestArguments;
+
+    const jsonRpcRequest = getJsonRpcRequest(
+      this.#nextRequestId++,
+      method,
+      params,
+    );
+    const jsonRpcResponse = await this.#fetchJsonRpcResponse(jsonRpcRequest);
+
+    if (isFailedJsonRpcResponse(jsonRpcResponse)) {
+      const error = new ProviderError(jsonRpcResponse.error.code);
+      error.data = jsonRpcResponse.error.data;
+
+      // eslint-disable-next-line no-restricted-syntax -- allow throwing ProviderError
+      throw error;
+    }
+
+    // TODO: emit hardhat network events (hardhat_reset, evm_revert)
+
+    return jsonRpcResponse.result;
+  }
+
+  /**
+   * @deprecated
+   * Sends a JSON-RPC request. This method is present for backwards compatibility
+   * with the Legacy Provider API. Prefer using `request` instead.
+   *
+   * @param method The method name for the JSON-RPC request.
+   * @param params The parameters for the JSON-RPC request. This should be an
+   * array of values.
+   * @returns The `result` property of the successful JSON-RPC response.
+   * @throws {ProviderError} If the JSON-RPC response indicates a failure.
+   * @throws {HardhatError} with descriptor:
+   * - {@link HardhatError.ERRORS.NETWORK.INVALID_REQUEST_METHOD} if the
+   * method is not a string.
+   * - {@link HardhatError.ERRORS.NETWORK.INVALID_REQUEST_PARAMS} if the
+   * params are not an array.
+   * - {@link HardhatError.ERRORS.NETWORK.CONNECTION_REFUSED} if the
+   * connection is refused.
+   * - {@link HardhatError.ERRORS.NETWORK.NETWORK_TIMEOUT} if the request
+   * times out.
+   */
+  public send(
+    method: string,
+    params?: unknown[],
+  ): Promise<SuccessfulJsonRpcResponse["result"]> {
+    return this.request({ method, params });
+  }
+
+  /**
+   * @deprecated
+   * Sends a JSON-RPC request asynchronously. This method is present for
+   * backwards compatibility with the Legacy Provider API. Prefer using
+   * `request` instead.
+   *
+   * @param jsonRpcRequest The JSON-RPC request object.
+   * @param callback The callback function to handle the response. The first
+   * argument should be an error object if an error occurred, and the second
+   * argument should be the JSON-RPC response object.
+   */
+  public sendAsync(
+    jsonRpcRequest: JsonRpcRequest,
+    callback: (error: any, response: JsonRpcResponse) => void,
+  ): void {
+    const handleJsonRpcRequest = async () => {
+      let jsonRpcResponse: JsonRpcResponse;
+      try {
+        const result = await this.request({
+          method: jsonRpcRequest.method,
+          params: jsonRpcRequest.params,
+        });
+        jsonRpcResponse = {
+          jsonrpc: "2.0",
+          id: jsonRpcRequest.id,
+          result,
+        };
+      } catch (error) {
+        ensureError<Error & { code: unknown }>(error);
```

Comment:
Note that using `ensureError<Foo>` like this is equivalent to casting, which can be incorrect at runtime.

This just happens to work because the first line after this throws if code is not defined, but this is a risky pattern.

Reply:
Done [here](https://github.com/NomicFoundation/hardhat/pull/5580/commits/7229bd0a281274f58361b6468a5957a5d29ca124)